perf: reorderOnSort option for collectionView

ptbrowne opened the pull request:
## Why

When the collectionView need to resort, the default implementation is to re-render. This can have performance problems on large CollectionViews where ChildViews are a bit complex.

## How

ChildViews have already been rendered once so instead of re-rendering them, reuse them and only sort DOM nodes.

## Possible problems

ChildViews that knows their index and use it in their rendering will know be able to use this technique since they need to be re-rendered after sorting. This is why this performance improvement is not enabled by default.

See https://github.com/marionettejs/backbone.marionette/issues/1979

ptbrowne commented:
I am not quite happy with the name of the feature flag, tell me if you have any suggestion.

megawac commented:
I'm :-1: for this without some decent benchmarks as you are still detaching all the elements in the tree and then readding them in proper order. If you were diff swapping nodes I could see how that would make a difference in some cases, but this method just avoids some of the render overhead.

ptbrowne commented:
I am all for doing benchmark if you like it but I could not find the jsLitmus files that were there before ? How can I include a benchmark ?

You can see http://jsbin.com/vecobada/13/edit?html,output for an example. Try with the option checked and without check. On my MBPr, the difference is cleary noticeable.

For what it’s worth, here is Chrome profile output of my example where I click on sort.

Here without the reorder option:
![screen shot 2015-02-07 at 16 53 15](https://cloud.githubusercontent.com/assets/465582/6092701/107e1a14-aeea-11e4-9d6e-c4cda34f0c6d.png)

Here with the reorder option (same scale):
![screen shot 2015-02-07 at 16 53 29](https://cloud.githubusercontent.com/assets/465582/6092702/1f4457d4-aeea-11e4-9758-6d97305e9856.png)

“just avoids some of the render overhead” -> If you have moderately complex elements, this render overhead can make all the difference between “laggy” and “instant”.

samccone commented:
yep @ptbrowne I am all for this, I think regardless of the final "optimized" implementation this is an amazing first step.

As far as perf, I think we can do something on jsperf. It will just require someone to run grunt build on this branch and use that for one test and the normal dist for the other.

:clap: for this great work so far

megawac commented on src/collection-view.js:
```diff
@@ -128,12 +128,36 @@ Marionette.CollectionView = Marionette.View.extend({
+    // transform into a jQuery array
+    var allEls = $(arr).map(function () { return this.toArray(); });
+    allEls.detach();
+    this.$el.append(allEls);
```
Correct me if there is a gotcha here, but I'm pretty sure this can just be

```diff
-    // transform into a jQuery array
-    var allEls = $(arr).map(function () { return this.toArray(); });
-    allEls.detach();
-    this.$el.append(allEls);
+    this.$el.append(arr);
```

ptbrowne replied:
Seems to work, I do not remember why I wrote it in this convoluted way :+1:

ptbrowne commented:
Another thing: If you have images in your ChildViews, they will be refetched if you re-render (normally the cache will take care of that but still, it is some work for the browser). When reordering nodes, they are not re-fetched.

http://jsbin.com/soxume/1/edit?html,output (check the network tab of Chrome Developer Tools)

cmaher commented on src/collection-view.js:
```diff
@@ -128,12 +128,34 @@ Marionette.CollectionView = Marionette.View.extend({
+    // get the views in the same order as the models
+    var arr = this.collection.models.map(function (model) {
+      var viewid = idx[model.cid];
+      var view = views[viewid];
```
`children.findByModel` would simplify this code

ptbrowne replied:
you're right :)

cmaher commented on test/unit/sorted-views.spec.js:
```diff
@@ -391,4 +391,30 @@ describe('collection/composite view sorting', function() {
+      this.collectionView = new this.CollectionView({
+        childView: this.ChildView,
+        collection: this.collection,
+        reorderOnSort: true
```
testing that `reorder` is called when `reorderOnSort` is set to true on the prototype would be good too.

ptbrowne replied:
`+      expect(this.collectionView.reorder).to.have.been.calledOnce;`

is already there. Did you mean something else ?

megawac replied:
I think @cmaher means setting `MArionette.CollectionView.prototype.reorderOnSort = true;`

samccone replied:
yeah so 

you would make a new collectionView extension

```js
MyView = Mn.CollectionView.extend({
  reorderOnSort: true
});
```
```js
new MyView({collection: this.collection})
```

ptbrowne replied:
understood :) thanks

cmaher commented:
This is looking pretty good. Some docs would also be nice.

samccone commented on src/collection-view.js:
```diff
@@ -128,12 +128,33 @@ Marionette.CollectionView = Marionette.View.extend({
+    var children = this.children;
+
+    // get the DOM nodes in the same order as the models
+    var els = this.collection.models.map(function (model) {
```
If we really want to be fancy

```js
_(this.collection.models)
.map(children.findByModel)
.pick(el)
.value()
```

ptbrowne replied:
I guess it would have to be `.map(this.children.findByModel, this.children)` and `.pick('el')` ;-) Anyway, I think until `underscore` becomes lazy, the chaining is uglified by the `.value()` at the end (that I always forgot). Also, I try to avoid `_`'s chaining because it is too easy to shoot yourself in the foot with the `this`.

samccone replied:
hmm not sure if it would be slower or not.  above is a more functional approach ^ that required less internal state. 
but not a real big deal one way or the other

megawac replied:
It would be slower but I wouldn't expect it to be an important issue

samccone commented on src/collection-view.js:
```diff
@@ -128,12 +128,33 @@ Marionette.CollectionView = Marionette.View.extend({
+      return children.findByModel(model).el;
+    });
+
+    // since append moves elements that are already in the DOM,
```
this is a cool behavior that i did not know about!

samccone commented on test/unit/sorted-views.spec.js:
```diff
@@ -391,4 +391,53 @@ describe('collection/composite view sorting', function() {
+      onPrototype: 'on the prototype'
+    };
+
+    var describeSpec = function (instance) {
```
#meta spec :pager:

samccone commented:
@ptbrowne thanks for the work! sorry we are getting into the tedious part of a PR of this nature. thanks for sticking it out!

ptbrowne commented:
No problem, it's normal. Happy to contribute. What do you mean by :pager: ?

samccone commented:
oh nothing, just a meta spec :) nothing wrong with it

samccone commented:
@ptbrowne on your next pass would you mind squashing this down to a single commit?

ptbrowne commented:
OK, no problem. How do I do that ? I squash on my machine and then `push -f` ?

samccone commented:
you got it!

samccone commented:
:+1: on this

paulfalgout commented:
So thoughts on adding events?  A normal resort would trigger `render` events.
`this.triggerMethod('before:reorder', this);`
`this.triggerMethod('reorder', this);`

Also this wouldn't work until after a `render` occurs.. it's probably ok because the only thing directly calling this function is a `sort` event that is setup after the first `render`.  Not sure if we need to handle or indicate that somewhere?

Also in the comments for `CollectionView`  sort and comparator are mentioned up at the constructor.  Do we want to keep those option comments together?

samccone commented:
ah good call on the events @paulfalgout , I think since it is a new "path" it would be wise to do this..

ptbrowne commented:
I agree with @paulfalgout and I added the events. Updated the docs too.

About the need for docs about the need to specify that a first render call is necessary, I think that it is both unnecessary since as you said the first `sort` event is sent after the first `render`. And also because in the docs, it is already stated that this option is useful when `resorting` the collection, and that it uses pre-existing DOM nodes. At least I can't find a formulation that would be clear and not confusing apart from what is already stated.

Thanks for the comments :)

ptbrowne commented:
Mmh I have just seen this https://github.com/marionettejs/backbone.marionette/commit/7fcdc351099b5b54edf32a2c4aa3097dc3c95b7f, it seems that `reorder` should support `viewComparator` right ?

samccone commented:
yep thats correct @ptbrowne , should be as simple as calling into `_sortedModels`

samccone commented:
@marionettejs/marionette-core one more set of eyes?

this is :+1: for me

cmaher commented on test/unit/sorted-views.spec.js:
```diff
@@ -391,4 +391,85 @@ describe('collection/composite view sorting', function() {
+      viewComparator: 'with viewComparator'
+    };
+
+    var getSpecTitle = function(options) {
```
is there a reason you're passing an object instead of an array?

ptbrowne replied:
`getSpecTitle` gets an object because the `specOptions` is an object meant to imitate a `set`. I could do `getSpecTitle(_.keys(specOptions))` and have:
```
getSpecTitle: function (options) {
   return _.map(options, function (k) { return texts[k]; })
}
```

but I do not think its better. Anyway, at this point, it really is a matter of choice.

cmaher commented on src/collection-view.js:
```diff
@@ -128,12 +128,39 @@ Marionette.CollectionView = Marionette.View.extend({
+    var viewComparator = Marionette.getOption(this, 'viewComparator');
+    var models = this.collection.models;
+    var mapper = (viewComparator ?
+      _.partial(_.map, this._sortedModels()) : _.bind(models.map, models));
```
`models.map` will break IE8

samccone replied:
dope great catch totally missed that!

ptbrowne replied:
perfect, it simplifies the code :)

cmaher commented on test/unit/sorted-views.spec.js:
```diff
@@ -391,4 +391,85 @@ describe('collection/composite view sorting', function() {
+            var ReorderedCollectionView = this.CollectionView.extend({
+              reorderOnSort: true,
+              onReorder: sinon.spy(),
+              onBeforeReorder: sinon.spy()
```
`this.sinon`

ptbrowne replied:
ok. Is there a reason why you do not use the global `sinon` object ? Did you have some problems with it being polluted during the tests ?

cmaher commented on test/unit/sorted-views.spec.js:
```diff
@@ -391,4 +391,85 @@ describe('collection/composite view sorting', function() {
+            expect(cv.onBeforeReorder).calledBefore(cv.onReorder);
+          }
+          expect(cv.trigger).to.have.been.calledWith('before:reorder');
+          expect(cv.trigger).to.have.been.calledWith('reorder');
```
`.and.calledOnce`

ptbrowne replied:
`.and.calledOnce` does not follow the “filter” on args done by `calledWith` so I put `         expect(cv.trigger).to.have.been.calledTwice;`

ptbrowne commented:
Travis failed `git clone git://github.com/jasonLaster/dox.git github.com[0: 192.30.252.129]: errno=Connection timed out`. Can someone restart the Travis job ?

jmeas commented:
Restarted

cmaher commented on test/unit/sorted-views.spec.js:
```diff
@@ -391,4 +391,86 @@ describe('collection/composite view sorting', function() {
+        });
+
+        it('should call reorder instead of render', function () {
+          expect(this.collectionView.render).not.to.have.been.calledOnce;
```
`not.to.have.been.called` is probably safer that `calledOnce` (in case it ever gets called twice)

ptbrowne replied:
good catch :)

cmaher commented:
One last tiny nit (sorry), and then I am :+1:  to this. It's looking great!

cmaher commented:
:+1: :shipit:

jmeas commented:
Nice work @ptbrowne!
